src/Validation/URLValidationProvider.php

```diff
@@ -0,0 +1,257 @@
+	 *
+	 * @return int
+	 */
+	public function get_unaccepted_errors() {
+		return $this->unaccepted_errors;
+	}
+
+	/**
+	 * Provides the number of URLs that have been checked.
+	 *
+	 * @return int
+	 */
+	public function get_number_validated() {
+		return $this->number_validated;
+	}
+
+	/**
+	 * Provides the validity counts by type.
+	 *
+	 * @return array[]
+	 */
+	public function get_validity_by_type() {
+		return $this->validity_by_type;
+	}
+
+	/**
+	 * Validates a URL, stores the results, and increments the counts.
+	 *
+	 * @param string $url  The URL to validate.
+	 * @param string $type The type of template, post, or taxonomy.
+	 * @param string $flag Flag determining whether the URL should be revalidated.
+	 * @return array|WP_Error Associative array containing validity result and whether the URL was revalidated, or a WP_Error on failure.
+	 */
+	public function get_url_validation( $url, $type, $flag = null ) {
+		$validity    = null;
+		$revalidated = true;
+
+		if ( self::FLAG_FORCE_REVALIDATE !== $flag ) {
+			$url_post = AMP_Validated_URL_Post_Type::get_invalid_url_post( $url );
+
+			if ( $url_post && empty( AMP_Validated_URL_Post_Type::get_post_staleness( $url_post ) ) ) {
+				$validity    = AMP_Validated_URL_Post_Type::get_invalid_url_validation_errors( $url_post );
+				$revalidated = false;
+			}
+		}
+
+		if ( self::FLAG_NO_REVALIDATE !== $flag && ( is_null( $validity ) || self::FLAG_FORCE_REVALIDATE === $flag ) ) {
+			$validity = AMP_Validation_Manager::validate_url_and_store( $url );
+		}
+
+		if ( is_wp_error( $validity ) ) {
+			return $validity;
+		}
+
+		if ( $validity && isset( $validity['results'] ) ) {
+			$this->update_state_from_validity( $validity, $type );
+		}
+
+		return compact( 'validity', 'revalidated' );
+	}
+
+	/**
+	 * Increments crawl counts from a validation result.
+	 *
+	 * @param array  $validity Validity results.
+	 * @param string $type The URL type.
+	 */
+	private function update_state_from_validity( $validity, $type ) {
+		$validation_errors      = wp_list_pluck( $validity['results'], 'error' );
+		$unaccepted_error_count = count(
+			array_filter(
+				$validation_errors,
+				static function( $error ) {
+					$validation_status = AMP_Validation_Error_Taxonomy::get_validation_error_sanitization( $error );
+					return (
+					AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_ACK_ACCEPTED_STATUS !== $validation_status['term_status']
+					&&
+					AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_NEW_ACCEPTED_STATUS !== $validation_status['term_status']
+					);
+				}
+			)
+		);
+
+		if ( count( $validation_errors ) > 0 ) {
+			$this->total_errors++;
+		}
+		if ( $unaccepted_error_count > 0 ) {
+			$this->unaccepted_errors++;
+		}
+
+		$this->number_validated++;
+
+		if ( ! isset( $this->validity_by_type[ $type ] ) ) {
+			$this->validity_by_type[ $type ] = [
+				'valid' => 0,
+				'total' => 0,
+			];
+		}
+			$this->validity_by_type[ $type ]['total']++;
```

Comment:
Strange that PHPCS did not catch this:

```suggestion
		$this->validity_by_type[ $type ]['total']++;
```